logic additions/fixes

foxsouns opened the pull request:
fisherman:
- casual is the now vanilla strat (`AND(FEATHER, FLIPPERS, TRADING_ITEM_FISHING_HOOK)`)
- normal is now `OR(FEATHER, SWORD, BOW), FLIPPERS, TRADING_ITEM_FISHING_HOOK` (due to the ease of doing so with sword/bow being next-to-vanilla)
- hard's method has been rediscovered, and is now documented
- glitched has a more informative comment
- removed from hell logic

richard's bushfield:
- logically requires `AND(MAX_POWDER_UPGRADE, POWDER)` instead of just powder, in order to prevent the rando asking the player to cross the whole field with only 10 powder

d6 connector:
- glitched can jesus jump over water with feather

d4/below taltal area:
- new hell jesus jumps: waterfalls one screen right of d4 ledge <-> staircase below damp cave
- hell d4 <-> staircase below damp cave jesus jump comment edit

foxsouns commented:
okay, actually, turning this into a logic-general PR. fisher featherless (hard) requires sword, or bombs (glitched). i got a seed that tried to do fisher without sword or bombs again: spent like 10 minutes trying with up held, and even tried tas tools (slowdown and turbo) to try and align it, and simply could not get it to work without sword. updating logic to match that momentarily.

foxsouns commented:
TODAY I LEARNED: changing a branch name is a `mv` action, not a rename action. damnit, git.

foxsouns commented:
OH! pulled fisher off with boomerang!! going to try a couple more like-items to see if i can get something else to work too.

foxsouns commented:
boomerang seems to be either y-pixel or y-subpixel perfect, and frame-perfect with the boat at its lowest, and is extremely tedious to get to work. the rang has to exit the water (which is inconsistent as hell already), and i think im going to throw that one into hell logic. however: bow works extremely consistently! that one is going in hard logic along with sword.

foxsouns commented:
i have a feeling that rod will work as well as sword, but i need to get it and test it first. mushroom/ocarina, hookshot, powder, and itemless are either as consistent as boomerang, or impossible. the others are items i have not tested.

foxsouns commented:
another logic hole spotted: seashell mansion bush does not allow boomerang, as you cannot get any angle where boomerang can come back over the bush to destroy it. might be possible, but id rather not force a "might", even on hell players.

foxsouns commented:
still need to test all items with fisher: i dont have a completed save file handy, i'll get around to it later today.

foxsouns commented:
> another logic hole spotted: seashell mansion bush does not allow boomerang, as you cannot get any angle where boomerang can come back over the bush to destroy it. might be possible, but id rather not force a "might", even on hell players.

found a working angle: spam under the bush at the top right. throwing this under hard logic.

kbranch commented:
My experience with boomeranging bushes like that has been that just spamming boomerang directly at them breaks them pretty quickly. It bounces off a few times but works relatively quickly. Does that not work here?

foxsouns commented:
> My experience with boomeranging bushes like that has been that just spamming boomerang directly at them breaks them pretty quickly. It bounces off a few times but works relatively quickly. Does that not work here?

it did, actually; i just pushed code for that in hard logic. still found it annoying, and would rather keep the possibility of requiring boomerang to those who play on hard & up

kbranch commented:
Maybe I misunderstood, I interpreted that as kind of a setup to get the return stroke to break the bush. This is what I had in mind, just to be clear - it does not seem to be sensitive to position:

![bgb_ct6sxCp7vX](https://user-images.githubusercontent.com/6884577/211064974-782a91db-be76-4ede-a161-095f08775fd7.gif)

I haven't done rigorous testing, but in my experience all recalcitrant bushes act like this. It might take a couple tries, but they all break without any special setup or input.

foxsouns commented:
sigh, bushes in this game are beyond weird. going to revert that to r.bush, i guess.

kbranch commented:
Yep, they're definitely weird. I had to try a few times to get a copy of that GIF that I liked. The first several times I hit record and loaded the state, the bush broke on the first try. I'm not sure what exactly happens behind the scenes, but it seems like there's a random chance of any given hit breaking the bush.

foxsouns commented:
@daid, i think i am FINALLY done with this. 
fisher: cas is the expected strat, normal is cas + sword or bow, hard is now explained, glitched has a better comment, removed from hell
richard's bushfield is also changed to require max powder instead of just powder, to prevent asking the player to cross the whole thing with only 10 powder.

foxsouns commented:
(you should probably merge as one commit, using the PR initial comment as the commit description)

foxsouns commented:
ONE SECOND:
found another thing: path to d6 is traversable in glitched logic with feather (one pit buffer jump, and one jesus jump).

foxsouns commented on logic/overworld.py:
```diff
@@ -209,7 +209,10 @@ def __init__(self, options, world_setup, r):
         bay_water.connect(left_bay_area, FLIPPERS)
         fisher_under_bridge = Location().add(TradeSequenceItem(0x2F5, TRADING_ITEM_NECKLACE))
-        fisher_under_bridge.connect(bay_water, AND(TRADING_ITEM_FISHING_HOOK, FEATHER, FLIPPERS))
+        if options.logic == "casual":
+            fisher_under_bridge.connect(bay_water, AND(TRADING_ITEM_FISHING_HOOK, FEATHER, FLIPPERS))
+        else
+            fisher_under_bridge.connect(bay_water, AND(TRADING_ITEM_FISHING_HOOK, OR(FEATHER, SWORD, BOW), FLIPPERS)) # just swing/shoot at fisher, trivially easy
```
@daid should i take this part out and just use `AND(TRADING_ITEM_FISHING_HOOK, FEATHER, FLIPPERS)`? i think sword/bow are dirt easy to do, and the differentiation between doing itemless is nice, but this isn't exactly "intended behavior" for normal logic, in retrospect, so i thought i should bring it up.

daid replied:
so, how does this work with bow? I'm not sure about this one yet, I've seen that you have been experimenting, but I haven't completely followed it (with me being sick and all)

foxsouns replied:
> so, how does this work with bow? I'm not sure about this one yet, I've seen that you have been experimenting, but I haven't completely followed it (with me being sick and all)

When you fire off a shot, for whatever reason, link's hitbox for triggers expands in front of him, and it makes it really easy to talk to the fisherman. I figure that it's due to having to check for bomb arrows. I get it first or second try with the bow: you just have to go directly underneath the fisherman, and shoot the bow while it's bound to the A button.

foxsouns commented:
oh.. pause again, i found some more tech.
the waterfalls from taltal can be jumped off of without flippers (by using s+q menu after ledge to regain pauses, and then 4 screens of jesus jumps) to get to the mountain area section. that definitely belongs in hell logic, though, no way am i going to require more than 1-2 screens of jesus jumps in glitched, lmao.
